/*
 *  Copyright (c) 2014, Lukas Tenbrink.
 *  * http://lukas.axxim.net
 */

package ivorius.psychedelicraft.block;

import java.util.Map;
import com.mojang.serialization.MapCodec;

import ivorius.psychedelicraft.block.entity.BarrelBlockEntity;
import ivorius.psychedelicraft.block.entity.PSBlockEntities;
import ivorius.psychedelicraft.fluid.*;
import ivorius.psychedelicraft.fluid.container.Resovoir;
import ivorius.psychedelicraft.item.component.FluidCapacity;
import ivorius.psychedelicraft.item.component.ItemFluids;
import ivorius.psychedelicraft.screen.FluidContraptionScreenHandler;
import ivorius.psychedelicraft.screen.PSScreenHandlers;
import net.minecraft.block.*;
import net.minecraft.block.entity.*;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.item.ItemPlacementContext;
import net.minecraft.item.ItemStack;
import net.minecraft.item.Items;
import net.minecraft.screen.ScreenHandlerType;
import net.minecraft.sound.SoundCategory;
import net.minecraft.sound.SoundEvents;
import net.minecraft.state.StateManager;
import net.minecraft.state.property.BooleanProperty;
import net.minecraft.state.property.EnumProperty;
import net.minecraft.state.property.Properties;
import net.minecraft.util.*;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Direction;
import net.minecraft.util.math.Direction.Axis;
import net.minecraft.util.shape.VoxelShape;
import net.minecraft.util.shape.VoxelShapes;
import net.minecraft.world.BlockView;
import net.minecraft.world.World;
import net.minecraft.world.block.WireOrientation;
import net.minecraft.world.event.GameEvent;

public class BarrelBlock extends FluidMachineBlock<BarrelBlockEntity> {
    public static final MapCodec<BarrelBlock> CODEC = createCodec(BarrelBlock::new);
    public static final int MAX_TAP_AMOUNT = FluidVolumes.BUCKET;
    public static final EnumProperty<Direction> FACING = Properties.HOPPER_FACING;
    public static final BooleanProperty TAPPED = BooleanProperty.of("tapped");

    private static final Map<Axis, VoxelShape> STANDING_SHAPES = Map.of(
        Axis.X, VoxelShapes.union(
            Block.createCuboidShape(0, 5, 2, 16, 13, 14),
            Block.createCuboidShape(0, 3, 4, 16, 15, 12)
        ),
        Axis.Y, VoxelShapes.union(
            Block.createCuboidShape(2, 0, 4, 14, 16, 12),
            Block.createCuboidShape(4, 0, 2, 12, 16, 14)
        ),
        Axis.Z, VoxelShapes.union(
            Block.createCuboidShape(2, 5, 0, 14, 13, 16),
            Block.createCuboidShape(4, 3, 0, 12, 15, 16)
        )
    );

    public BarrelBlock(Settings settings) {
        super(settings.nonOpaque());
        setDefaultState(getDefaultState().with(FACING, Direction.NORTH).with(TAPPED, true));
    }

    @Override
    protected MapCodec<? extends BarrelBlock> getCodec() {
        return CODEC;
    }

    @Override
    protected void appendProperties(StateManager.Builder<Block, BlockState> builder) {
        super.appendProperties(builder);
        builder.add(FACING, TAPPED);
    }

    @Override
    protected BlockEntityType<BarrelBlockEntity> getBlockEntityType() {
        return PSBlockEntities.BARREL;
    }

    @Override
    protected ScreenHandlerType<FluidContraptionScreenHandler<BarrelBlockEntity>> getScreenHandlerType() {
        return PSScreenHandlers.BARREL;
    }

    @Override
    protected BlockRenderType getRenderType(BlockState state) {
        return BlockRenderType.INVISIBLE;
    }

    @Override
    protected VoxelShape getOutlineShape(BlockState state, BlockView world, BlockPos pos, ShapeContext context) {
        return STANDING_SHAPES.get(state.get(FACING).getAxis());
    }

    @Override
    protected BlockState rotate(BlockState state, BlockRotation rotation) {
        return state.with(FACING, rotation.rotate(state.get(FACING)));
    }

    @Override
    protected BlockState mirror(BlockState state, BlockMirror mirror) {
        return state.rotate(mirror.getRotation(state.get(FACING)));
    }

    @Override
    public BlockState getPlacementState(ItemPlacementContext ctx) {
        Direction side = ctx.getPlayerLookDirection().getOpposite();
        return getDefaultState().with(FACING, side.getAxis() == Axis.Y ? Direction.DOWN : side);
    }

    @Override
    protected ActionResult onInteractWithItem(ItemStack stack, BlockState state, World world, BlockPos pos, PlayerEntity player, Hand hand, BarrelBlockEntity blockEntity) {

        if (stack.isOf(Items.STICK)) {
            world.playSound(player, pos, SoundEvents.BLOCK_BARREL_CLOSE, SoundCategory.BLOCKS);
            world.emitGameEvent(player, GameEvent.BLOCK_OPEN, pos);
            world.setBlockState(pos, state.with(FACING, state.get(FACING).getAxis() == Axis.Y ? player.getHorizontalFacing().getOpposite() : Direction.DOWN));
            return ActionResult.SUCCESS;
        }

        int capacity = FluidCapacity.get(stack);
        if (!state.get(TAPPED) || state.get(FACING).getAxis() == Axis.Y || capacity == 0) {
            return ActionResult.PASS_TO_DEFAULT_BLOCK_ACTION;
        }

        if (ItemFluids.of(stack).amount() < capacity) {
            Resovoir tank = blockEntity.getTankOnSide(Direction.DOWN);
            if (tank.getContents().amount() > 0 && stack.isIn(tank.getContents().fluid().getPreferredContainerTag())) {
                if (!world.isClient) {

                    ItemFluids.Transaction t = ItemFluids.Transaction.begin(stack.copyWithCount(1));

                    if (tank.withdraw(t, MAX_TAP_AMOUNT) > 0) {
                        if (stack.getCount() > 1) {
                            stack.decrement(1);
                            player.getInventory().offerOrDrop(t.toItemStack());
                        } else {
                            player.setStackInHand(hand, t.toItemStack());
                        }
                    }

                    blockEntity.setTapOpenTicks(20);
                    blockEntity.markForUpdate();

                    Direction updateDirection = state.get(BarrelBlock.FACING).getOpposite();
                    if (updateDirection.getAxis() != Axis.Y) {
                        world.updateNeighborsExcept(pos.offset(updateDirection), this, updateDirection.getOpposite(), WireOrientation.random(world.random));
                    }
                }

                return ActionResult.SUCCESS;
            }
        }

        return ActionResult.FAIL;
    }

    @Override
    protected boolean emitsRedstonePower(BlockState state) {
        return state.get(FACING).getAxis() != Axis.Y;
    }

    @Override
    protected int getWeakRedstonePower(BlockState state, BlockView world, BlockPos pos, Direction direction) {
        // only output at the back
        if (direction != state.get(FACING)) {
            return 0;
        }
        // signal 0-15 to indicate progress
        return super.getWeakRedstonePower(state, world, pos, direction);
    }

    @Override
    protected int getStrongRedstonePower(BlockState state, BlockView world, BlockPos pos, Direction direction) {
        // only output at the back
        if (direction != state.get(FACING)) {
            return 0;
        }
        // signal 15 = Someone is using it
        if (world.getBlockEntity(pos, getBlockEntityType()).map(BarrelBlockEntity::getTapOpenTicks).orElse(0) > 0) {
            return 15;
        }
        return 0;
    }
}
